plugins/chat/spec/services/chat/auto_join_channel_batch_spec.rb

```diff
@@ -0,0 +1,281 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+describe Chat::AutoJoinChannelBatch do
+  subject(:result) { described_class.call(params) }
+
+  let(:chat_channel_id) { channel_1.id }
+  let(:starts_at) { 0 }
+  let(:ends_at) { 10 }
+  let(:params) { { chat_channel_id: chat_channel_id, starts_at: starts_at, ends_at: ends_at } }
+
+  fab!(:channel_1) { Fabricate(:chat_channel, auto_join_users: true) }
+
+  context "when contract" do
```

Comment:
I’m just missing specs for the new action, I’m putting what I have already here:

_spec/services/chat/auto_join_channel_batch_spec.rb_
```ruby
# frozen_string_literal: true

require "rails_helper"

describe Chat::AutoJoinChannelBatch do
  describe Chat::AutoJoinChannelBatch::Contract, type: :model do
    subject(:contract) { described_class.new(start_user_id: 10) }

    it { is_expected.to validate_presence_of(:channel_id) }
    it { is_expected.to validate_presence_of(:start_user_id) }
    it { is_expected.to validate_presence_of(:end_user_id) }
    it { is_expected.to validate_comparison_of(:end_user_id).is_greater_than(:start_user_id) }

    describe "Backward compatibility" do
      subject(:contract) { described_class.new(args) }

      before { contract.valid? }

      context "when providing 'chat_channel_id'" do
        let(:args) { { chat_channel_id: 2 } }

        it "sets 'channel_id'" do
          expect(contract.channel_id).to eq(2)
        end
      end

      context "when providing 'starts_at'" do
        let(:args) { { starts_at: 5 } }

        it "sets 'start_user_id'" do
          expect(contract.start_user_id).to eq(5)
        end
      end

      context "when providing 'ends_at'" do
        let(:args) { { ends_at: 8 } }

        it "sets 'end_user_id'" do
          expect(contract.end_user_id).to eq(8)
        end
      end
    end
  end

  describe ".call" do
    subject(:result) { described_class.call(params) }

    fab!(:channel) { Fabricate(:chat_channel, auto_join_users: true) }

    let(:channel_id) { channel.id }
    let(:start_user_id) { 0 }
    let(:end_user_id) { 10 }
    let(:params) do
      { channel_id: channel_id, start_user_id: start_user_id, end_user_id: end_user_id }
    end

    context "when arguments are invalid" do
      let(:channel_id) { nil }

      it { is_expected.to fail_a_contract }
    end

    context "when arguments are valid" do
      context "when channel does not exist" do
        let(:channel_id) { -1 }

        it { is_expected.to fail_to_find_a_model(:channel) }
      end

      context "when channel is not a category channel" do
        fab!(:channel) { Fabricate(:direct_message_channel, auto_join_users: true) }

        it { is_expected.to fail_to_find_a_model(:channel) }
      end

      context "when channel is not in auto_join_users mode" do
        before { channel.update!(auto_join_users: false) }

        it { is_expected.to fail_to_find_a_model(:channel) }
      end

      context "when channel is found" do
        fab!(:users) { Fabricate.times(2, :user) }

        let(:manager) { mock.responds_like_instance_of(Chat::ChannelMembershipManager) }

        before do
          Chat::Action::CreateMembershipsForAutoJoin
            .stubs(:call)
            .with(has_entries(channel: channel, contract: instance_of(described_class::Contract)))
            .returns(user_ids)
          Chat::ChannelMembershipManager.stubs(:new).with(channel).returns(manager)
          manager.stubs(:recalculate_user_count)
        end

        context "when more than one membership is created" do
          let(:user_ids) { users.map(&:id) }

          it "does not recalculate user count" do
            manager.expects(:recalculate_user_count).never
            result
          end

          it "publishes an event" do
            Chat::Publisher.expects(:publish_new_channel).with(channel, users)
            result
          end
        end

        context "when only one membership is created" do
          let(:user_ids) { [users.first.id] }

          it "recalculates user count" do
            manager.expects(:recalculate_user_count)
            result
          end

          it "publishes an event" do
            Chat::Publisher.expects(:publish_new_channel).with(channel, [users.first])
            result
          end
        end
      end
    end
  end
end
```
_app/services/chat/auto_join_channel_batch.rb_
```ruby
# NOTE: When changing auto-join logic, make sure to update the `settings.auto_join_users_info` translation as well.
# frozen_string_literal: true

module Chat
  # Service responsible to create memberships for a channel and a section of user ids
  #
  # @example
  #  Chat::AutoJoinChannelBatch.call(
  #   channel_id: 1,
  #   start_user_id: 27,
  #   end_user_id: 58,
  #  )
  #
  class AutoJoinChannelBatch
    include Service::Base

    contract
    model :channel
    step :create_memberships
    step :recalculate_user_count
    step :publish_new_channel

    class Contract
      # Backward-compatible attributes
      attribute :chat_channel_id, :integer
      attribute :starts_at, :integer
      attribute :ends_at, :integer

      # New attributes
      attribute :channel_id, :integer
      attribute :start_user_id, :integer
      attribute :end_user_id, :integer

      validates :channel_id, :start_user_id, :end_user_id, presence: true
      validates :end_user_id, comparison: { greater_than: :start_user_id }

      # TODO: remove after migration is done
      before_validation do
        self.channel_id ||= chat_channel_id
        self.start_user_id ||= starts_at
        self.end_user_id ||= ends_at
      end
    end

    private

    def fetch_channel(contract:, **)
      ::Chat::CategoryChannel.find_by(id: contract.channel_id, auto_join_users: true)
    end

    def create_memberships(channel:, contract:, **)
      context.added_user_ids =
        Chat::Action::CreateMembershipsForAutoJoin.call(channel: channel, contract: contract)
    end

    def recalculate_user_count(channel:, added_user_ids:, **)
      # Only do this if we are running auto-join for a single user, if we
      # are doing it for many then we should do it after all batches are
      # complete for the channel in Jobs::AutoJoinChannelMemberships
      return unless added_user_ids.one?
      ::Chat::ChannelMembershipManager.new(channel).recalculate_user_count
    end

    def publish_new_channel(channel:, added_user_ids:, **)
      ::Chat::Publisher.publish_new_channel(channel.reload, User.where(id: added_user_ids))
    end
  end
end
```
_app/services/chat/action/create_memberships_for_auto_join.rb_
```ruby
# frozen_string_literal: true

module Chat
  module Action
    class CreateMembershipsForAutoJoin
      def self.call(channel:, contract:)
        query_args = {
          chat_channel_id: channel.id,
          start: contract.start_user_id,
          end: contract.end_user_id,
          suspended_until: Time.zone.now,
          last_seen_at: 3.months.ago,
          channel_category: channel.category.id,
          permission_type: CategoryGroup.permission_types[:create_post],
          everyone: Group::AUTO_GROUPS[:everyone],
          mode: ::Chat::UserChatChannelMembership.join_modes[:automatic],
        }
        ::DB.query_single(<<~SQL, query_args)
          INSERT INTO user_chat_channel_memberships (user_id, chat_channel_id, following, created_at, updated_at, join_mode)
          SELECT DISTINCT(users.id), :chat_channel_id, TRUE, NOW(), NOW(), :mode
          FROM users
          INNER JOIN user_options uo ON uo.user_id = users.id
          LEFT OUTER JOIN user_chat_channel_memberships uccm ON
            uccm.chat_channel_id = :chat_channel_id AND uccm.user_id = users.id

          LEFT OUTER JOIN group_users gu ON gu.user_id = users.id
          LEFT OUTER JOIN category_groups cg ON cg.group_id = gu.group_id AND
          cg.permission_type <= :permission_type

          WHERE (users.id >= :start AND users.id <= :end) AND
            users.staged IS FALSE AND
            users.active AND
            NOT EXISTS(SELECT 1 FROM anonymous_users a WHERE a.user_id = users.id) AND
            (suspended_till IS NULL OR suspended_till <= :suspended_until) AND
            (last_seen_at IS NULL OR last_seen_at > :last_seen_at) AND
            uo.chat_enabled AND

            (NOT EXISTS(SELECT 1 FROM category_groups WHERE category_id = :channel_category)
              OR EXISTS (SELECT 1 FROM category_groups WHERE category_id = :channel_category AND group_id = :everyone AND permission_type <= :permission_type)
              OR cg.category_id = :channel_category)

          ON CONFLICT DO NOTHING

          RETURNING user_chat_channel_memberships.user_id
        SQL
      end
    end
  end
end
```

Reply:
Yep nice :+1: I considered writing an action for the insert too but decided against it for now,, but agreed it's the best :+1: and now that's it's done 😅

Reply:
And here is the spec for the new action (mostly it’s just what you wrote previously):
```ruby
# frozen_string_literal: true

RSpec.describe Chat::Action::CreateMembershipsForAutoJoin do
  subject(:action) { described_class.call(channel: channel, contract: contract) }

  fab!(:channel) { Fabricate(:chat_channel, auto_join_users: true) }
  fab!(:user_1) { Fabricate(:user, last_seen_at: 15.minutes.ago) }

  let(:start_user_id) { user_1.id }
  let(:end_user_id) { user_1.id }
  let(:contract) { OpenStruct.new(start_user_id: start_user_id, end_user_id: end_user_id) }

  it "adds correct members" do
    expect(action).to eq([user_1.id])
  end

  it "sets the reason to automatic" do
    action
    expect(channel.membership_for(user_1)).to be_automatic
  end

  context "with others users not in the batch" do
    fab!(:user_2) { Fabricate(:user) }

    it "adds correct members" do
      expect(action).to eq([user_1.id])
    end
  end

  context "with suspended users" do
    before { user_1.update!(suspended_till: 1.year.from_now) }

    it "skips suspended users" do
      expect(action).to eq([])
    end
  end

  context "with users not seen recently" do
    before { user_1.update!(last_seen_at: 4.months.ago) }

    it "skips users last_seen more than 3 months ago" do
      expect(action).to eq([])
    end
  end

  context "with never seen users" do
    before { user_1.update!(last_seen_at: nil) }

    it "skips users with last_seen set to null" do
      expect(action).to eq([user_1.id])
    end
  end

  context "with disabled chat users" do
    before { user_1.user_option.update!(chat_enabled: false) }

    it "skips users without chat_enabled" do
      expect(action).to eq([])
    end
  end

  context "with anonymous users" do
    fab!(:user_1) { Fabricate(:anonymous, last_seen_at: 15.minutes.ago) }

    it "skips anonymous users" do
      expect(action).to eq([])
    end
  end

  context "with inactive users" do
    before { user_1.update!(active: false) }

    it "skips inactive users" do
      expect(action).to eq([])
    end
  end

  context "with staged users" do
    before { user_1.update!(staged: true) }

    it "skips staged users" do
      expect(action).to eq([])
    end
  end

  context "when user is already a member" do
    before { channel.add(user_1) }

    it "is a noop" do
      expect(action).to eq([])
    end
  end

  context "when category is restricted" do
    fab!(:user_1) { Fabricate(:user) }
    fab!(:user_2) { Fabricate(:user) }
    fab!(:group_1) { Fabricate(:group) }
    fab!(:channel) { Fabricate(:private_category_channel, group: group_1, auto_join_users: true) }

    let(:end_user_id) { user_2.id }

    before { group_1.add(user_1) }

    it "only joins users with access to the category through the group" do
      expect(action).to eq([user_1.id])
    end

    context "when the user has access through multiple groups" do
      fab!(:group_2) { Fabricate(:group) }

      before do
        channel.category.category_groups.create!(
          group_id: group_2.id,
          permission_type: CategoryGroup.permission_types[:full],
        )
        group_2.add(user_1)
      end

      it "correctly joins the user" do
        expect(action).to eq([user_1.id])
      end
    end

    context "when the category group is read only" do
      fab!(:channel) { Fabricate(:private_category_channel, auto_join_users: true) }

      before do
        channel.category.category_groups.create!(
          group_id: group_1.id,
          permission_type: CategoryGroup.permission_types[:readonly],
        )
        group_1.add(user_1)
      end

      it "doesn’t join the users of the group" do
        expect(action).to eq([])
      end
    end

    context "when the category group has create post permission" do
      fab!(:channel) { Fabricate(:private_category_channel, auto_join_users: true) }

      before do
        channel.category.category_groups.create!(
          group_id: group_1.id,
          permission_type: CategoryGroup.permission_types[:create_post],
        )
        group_1.add(user_1)
      end

      it "correctly joins the user" do
        expect(action).to eq([user_1.id])
      end
    end

    context "when user has allowed groups and disallowed groups" do
      fab!(:group_2) { Fabricate(:group) }

      before do
        channel.category.category_groups.create!(
          group_id: group_2.id,
          permission_type: CategoryGroup.permission_types[:readonly],
        )
        group_2.add(user_1)
      end

      it "correctly joins the user" do
        expect(action).to eq([user_1.id])
      end
    end
  end
end
```
There’s something weird though:
```ruby
  context "with never seen users" do
    before { user_1.update!(last_seen_at: nil) }

    it "skips users with last_seen set to null" do
      expect(action).to eq([user_1.id])
    end
  end
```
This one is green so it might be a bug? :thinking:

Reply:
Just remembered that we have to use the `main` branch of `shoulda-matchers` to have access to the `validate_comparison_of` matcher (I guess they will release a new version soon).